### PR TITLE
[ENH] dropna parameter for `pivot_longer`

### LINKED ISSUE #1132: Allow dropna for pivot_longer
This might be more performant than dropping after the transformation, and should be similar to what pandas stack provides

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- add `dropna` parameter to drop nulls , similar to stack
- improve speed for scenarios where dot value is not involved - close to melt speed
- improve speed in some cases for multiple values_to

**This PR resolves #1132.**

Tests ... pinch of salt

```py
url = 'https://raw.githubusercontent.com/tidyverse/tidyr/main/data-raw/billboard.csv'
df = pd.read_csv(url)
df = pd.concat([df]*100, ignore_index = True)
df.shape
(31700, 81)

 %timeit df.melt(['year', 'artist', 'track', 'time', 'date.entered'], ignore_index = False).dropna(subset=['value'])
232 ms ± 3.37 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit df.pivot_longer(column_names = 'wk*', dropna = True, ignore_index = True)
196 ms ± 2.56 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

A = df.melt(['year', 'artist', 'track', 'time', 'date.entered'], ignore_index = False).dropna(subset=['value'])
B = df.pivot_longer(column_names = 'wk*', dropna = True, ignore_index = True)

A.reset_index(drop=True).equals(B)
True

%timeit df.pivot_longer(column_names = 'wk*', dropna = False)
146 ms ± 2.62 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit df.melt(['year', 'artist', 'track', 'time', 'date.entered'])
154 ms ± 1.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.md`.

3. [x] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1136--pyjanitor.netlify.app

File: janitor/functions/pivot.py
Comment (thatlittleboy) [inline janitor/functions/pivot.py]: Is this intentional? Setting outcome=values and values=outcome

Comment (samukweku) [reply janitor/functions/pivot.py]: yes it is; maybe I could do a better naming here maybe? values are what you have after `melt`ing the data, outcome is/are the column labels. the processing in this section could be values only, or values and columns; the assignment to the _final...longer function just ensures the processing downstream is streamlined, especially when dropping nulls (so they can be dropped in a single location, - the values variable)

Comment (thatlittleboy) [reply janitor/functions/pivot.py]: I'll need to delve through the code again, it just caught my eye;  
because previously it was `values=values`, `outcome=outcome`, so was that a bug? I think I'm just missing why adding the `dropna` functionality would flip this relationship around -- I'm not seeing specifically which part of the newly introduced code would result in this specific behaviour.

Comment (samukweku) [reply janitor/functions/pivot.py]: Sure, I'll rework d code and add comments. Hopefully I don't make it worse 🤣

Comment (samukweku) [reply janitor/functions/pivot.py]: @thatlittleboy I renamed the variables; the logic remains the same. Kindly have a look when you can, and let me know where it still has kinks. thanks

**Review by ericmjl — APPROVED** [review]
I approve! Once we have two approvals, the 2nd person who approves can do the honours of merging :smile:.

**Review by thatlittleboy — APPROVED** [review]

**Pull request merged**
